tvm/src/pass/stream_inference.cc

```diff
@@ -21,32 +44,887 @@ inline Expr Type2Expr(const Type& t) {
   return StringImm::make(os.str());
 }
 
+// Substitute the target buffer consumers with channel buffers
+class NewChannelGathers final : public IRMutator {
+ public: 
+  NewChannelGathers(vector<int> _index_array,
+    string _target_buffer_name,
+    StreamInfo _target_buffer_stream_info,
+    unordered_map<int, VarExpr>&  _channel_index_to_new_buffers) :
+    index_array(_index_array), 
+    target_buffer_name(_target_buffer_name),
+    target_buffer_stream_info(_target_buffer_stream_info),
+    channel_index_to_new_buffers(_channel_index_to_new_buffers) {}
+
+  Stmt Mutate(Stmt stmt) final {
+    if (!hit_target_channel_load) {
+        Stmt ret = IRMutator::Mutate(stmt);
+
+        // Add temp to save value before the statement
+        if (hit_target_channel_load) {
+            if (search_first_stmt_with_target == 0) {
+                HCL_DEBUG_LEVEL(2) << "Insert streaming channel reader of "
+                    << target_buffer_name << " before "
+                    << "the first Stmt consumer: "<< ret;
+
+                // Loading data from the channel 
+                // TODO: support multiple index case
+                auto index = index_array[0]; 
+                auto target_load_op = target_load_expr.as<Load>();
+                CHECK(target_load_op);
+                CHECK(channel_index_to_new_buffers.count(index));
+                VarExpr channel_buf(channel_index_to_new_buffers[index].node_); 
+                Expr new_load = Load::make(target_load_op->type, 
+                    channel_buf, target_load_op->index, target_load_op->predicate);
+
+                Stmt s;
+                if (insert_conditional_load) {
+                    HCL_DEBUG_LEVEL(2) << "[debug] Hmm.. Insert cond store...";
+                    CHECK(condition_node.defined());
+                    auto select_op = condition_node.as<Select>();
+                    CHECK(select_op); 
+
+                    // Merge the two conditional stores
+                    if (auto store_op = ret.as<Store>()) {
+                        CHECK(store_op->value.as<Select>());
```

Comment:
This checking cannot pass something like `C[i] = (A[i] < B.pipe[i]) ? 1 : 0`.

Reply:
Fixed.

Reply:
Seems not totally fixed... I got this error now.
```
heterocl.tvm._ffi.base.TVMError: [10:20:28] src/pass/stream_inference.cc:206: Check failed: !buffer_created Failure: trying to stream a tensor that has been written for multiple times...
```

Reply:
If I comment out these few lines, my program is able to work.
https://github.com/cornell-zhang/heterocl/blob/b105236cf8bc0d21b5813d65d40cd8ffa1089436/tvm/src/pass/stream_inference.cc#L81-L101

So there must be some problems here.

Reply:
Do you have a test case for it? If it is hard to create a test case. We can just roll back to the old version first, where we generate code like:
```c++
B_temp = select(i < 10, B_pipe, 0)
C[i] = select((i< 10), B_temp, some expressions)
```